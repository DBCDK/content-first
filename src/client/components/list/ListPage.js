import React from 'react';
import {connect} from 'react-redux';
import Spinner from '../general/Spinner.component';
import Title from '../base/Title';
import SimpleList from './templates/simple/SimpleList';
import BookcaseList from './templates/bookcase/BookcaseList';
import {getListByIdSelector} from '../../redux/list.reducer';
import {LIST_LOAD_REQUEST} from '../../redux/list.reducer';

const getListById = getListByIdSelector();

export class ListPage extends React.Component {
  componentDidMount() {
    this.props.loadList(this.props.id);
  }
  componentDidUpdate(prevProps) {
    if (prevProps.id !== this.props.id) {
      this.props.loadList(this.props.id);
    }
  }

  // eslint-disable-next-line no-unused-vars
  getTemplate(list) {
    const template = list.template;

    switch (template) {
      case 'list':
        return SimpleList;
      case 'bookcase':
        return BookcaseList;
      default:
        return SimpleList;
    }
  }
  render() {
    console.log('ListPage render');
    const {list} = this.props;
    if (!list || list.isLoading) {
      // TODO make a skeleton view of list
      return (
        <div className="d-flex justify-content-center">
          <Spinner size="30px" className="mt-5" />
        </div>
      );
    }

    if (list.error) {
      return (
        <Title Tag="h1" type="title3" className="text-center mt-5">
          Listen kunne ikke hentes
        </Title>
      );
    }
    if (!list._type) {
      return (
        <Title Tag="h1" type="title3" className="text-center mt-5">
          Listen er slettet
        </Title>
      );
    }

    const Template = this.getTemplate(list);

    return <Template _id={list._id} />;
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    list: getListById(state, {_id: ownProps.id})
  };
};

export const mapDispatchToProps = dispatch => ({
  loadList: _id => dispatch({type: LIST_LOAD_REQUEST, _id})
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ListPage);
