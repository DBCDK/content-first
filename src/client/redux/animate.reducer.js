const defaultState = {};

export const ANIMATE_START = 'ANIMATE_START';
export const ANIMATE_STOP = 'ANIMATE_STOP';

const animateReducer = (state = defaultState, action) => {
  switch (action.type) {
    case ANIMATE_START: {
      if (!action.name) {
        throw new Error('! No animation name found');
      }
      if (!action.component) {
        throw new Error('! No component found');
      }
      if (!action.from || !action.to) {
        throw new Error('! You need to set both a "from" and "to" position');
      }

      console.log('ANIMATE_START');

      const newState = Object.assign({}, state, {
        ...state,
        [action.name]: {
          name: action.name,
          component: action.component,
          from: action.from,
          to: action.to,
          onEnd: action.onEnd,
          animating: true,
          animated: false
        }
      });

      return newState;
    }

    case ANIMATE_STOP: {
      if (!action.name) {
        throw new Error('! No animation name found');
      }

      console.log('ANIMATE_STOP');

      return Object.assign({}, state, {
        [action.name]: {
          animating: false,
          animated: true
        }
      });
    }

    default:
      return state;
  }
};

export const startAnimate = animate => {
  console.log('startAnimate action', animate);

  return {
    type: ANIMATE_START,
    name: animate.name,
    component: animate.component,
    from: animate.from,
    to: animate.to,
    onEnd: animate.onEnd
  };
};

export const stopAnimate = name => {
  console.log('stopAnimate name', name);

  return {
    type: ANIMATE_STOP,
    name
  };
};

export default animateReducer;
