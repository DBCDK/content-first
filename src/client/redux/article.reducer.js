/* eslint import/no-webpack-loader-syntax: off */
import About from '!raw-loader!../components/article/pages/about.md';
import NotFound from '!raw-loader!../components/article/pages/404.md';

const defaultState = {
  articles: {
    '/om': {
      id: 1,
      name: 'om',
      title: 'Om LÃ¦sekompasset',
      meta: {
        title: 'Om',
        canonical: '/om',
        og: {
          title: 'Om lasekompas.dk',
          description: 'Hvad er laesekompas.dk?',
          url: 'https://laesekompas.dk/om',
          type: 'website',
          image: {
            url: 'http://laesekompas.dk/img/open-graph/hero-01.jpg',
            width: '1200',
            height: '675'
          }
        }
      },
      src: About
    },
    '/404': {
      id: 2,
      name: '404',
      meta: {
        title: '404 - Siden blev ikke fundet',
        canonical: '/404',
        robots: 'noindex'
      },
      src: NotFound
    }
  },
  isLoading: false
};

const articleReducer = (state = defaultState, action) => {
  switch (action.type) {
    default:
      return state;
  }
};

export default articleReducer;
