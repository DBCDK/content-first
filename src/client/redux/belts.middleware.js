import {
  BELTS_LOAD_REQUEST,
  BELTS_LOAD_RESPONSE,
  STORE_BELT,
  UPDATE_BELT,
  REMOVE_BELT_REQUEST,
  REMOVE_BELT_RESPONSE
} from './belts.reducer';
import {
  fetchObjects,
  addObject,
  updateObject,
  deleteObject
} from '../utils/requester';

export const beltsMiddleware = store => next => async action => {
  switch (action.type) {
    case STORE_BELT: {
      const openplatformId = store.getState().userReducer.openplatformId;

      const response = await addObject({
        ...action.belt,
        _owner: openplatformId
      });

      const belt = {
        ...action.belt,
        ...response.data,
        _owner: openplatformId
      };

      return next({type: action.type, belt});
    }

    case UPDATE_BELT: {
      const openplatformId = store.getState().userReducer.openplatformId;

      const response = await updateObject({
        ...action.belt,
        editing: false,
        _owner: openplatformId
      });

      const belt = {
        ...action.belt,
        ...response.data
      };

      return next({type: action.type, belt});
    }

    case REMOVE_BELT_REQUEST: {
      const _id = action.belt._id;
      return (async () => {
        next(action);

        deleteObject({_id});

        store.dispatch({
          type: REMOVE_BELT_RESPONSE
        });
      })();
    }

    case BELTS_LOAD_REQUEST: {
      const openplatformId = store.getState().userReducer.openplatformId;
      return (async () => {
        next(action);
        const belts = (await fetchObjects(null, 'belt', openplatformId)).data;

        store.dispatch({
          type: BELTS_LOAD_RESPONSE,
          data: belts
        });
      })();
    }
    default:
      return next(action);
  }
};
