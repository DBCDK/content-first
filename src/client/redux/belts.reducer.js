const defaultState = {
  belts: [
    {
      name: 'En god bog',
      details: 'Detaljer for en god bog',
      isLoading: false,
      onFrontPage: true,
      links: ['Spændende', 'Anderledes'],
      scrollOffset: 0,
      works: []
    },
    {
      name: 'Let læst',
      isLoading: false,
      onFrontPage: true,
      links: ['Klog', 'Udfordrende'],
      scrollOffset: 0
    },
    {
      name: 'Bibliotekarens ugentlige anbefalinger',
      details: 'Detaljer for ugentlige anbefalinger',
      isLoading: false,
      onFrontPage: true,
      links: [],
      scrollOffset: 0,
      works: []
    },
    {
      name: 'Passer med min smag',
      details: 'Detaljer for min smag',
      isLoading: false,
      onFrontPage: true,
      links: []
    },
    {
      name: 'Spændende',
      isLoading: false,
      onFrontPage: false,
      links: []
    },
    {
      name: 'Anderledes',
      isLoading: false,
      onFrontPage: false,
      links: []
    },
    {
      name: 'Klog',
      isLoading: false,
      onFrontPage: false,
      links: []
    },
    {
      name: 'Udfordrende',
      isLoading: false,
      onFrontPage: false,
      links: []
    }
  ]
};

export const ON_BELT_REQUEST = 'ON_BELT_REQUEST';
export const ON_BELT_RESPONSE = 'ON_BELT_RESPONSE';
export const ON_BELT_SCROLL = 'ON_BELT_SCROLL';
export const ON_TAG_TOGGLE = 'ON_TAG_TOGGLE';

const beltsReducer = (state = defaultState, action) => {
  switch (action.type) {
    case ON_BELT_REQUEST: {
      const belts = state.belts.map(belt => {
        if (belt.name === action.beltName) {
          return Object.assign({}, belt, {isLoading: true});
        }
        return belt;
      });
      return Object.assign({}, {belts});
    }

    case ON_BELT_RESPONSE: {
      const belts = state.belts.map(belt => {
        if (belt.name === action.beltName) {
          return Object.assign({}, belt, {isLoading: false}, {works: action.response});
        }
        return belt;
      });
      return Object.assign({}, {belts});
    }

    case ON_BELT_SCROLL: {
      const belts = [...state.belts];
      const belt = belts[action.id];
      if (action.scrollOffset < 0) {
        action.scrollOffset = 0;
      }
      else if (action.scrollOffset > belt.works.length - 1) {
        action.scrollOffset = belt.works.length - 1;
      }
      belts[action.id] = Object.assign({}, belt, {scrollOffset: action.scrollOffset});
      return Object.assign({}, {belts});
    }

    case ON_TAG_TOGGLE: {
      const belts = [...state.belts];
      const belt = Object.assign({}, belts[action.beltId]);
      belts[action.beltId] = belt;
      belt.tags = [...belt.tags];
      belt.tags[action.tagId] = {
        name: belt.tags[action.tagId].name,
        selected: belt.tags[action.tagId].selected === false
      };
      return Object.assign({}, {belts});
    }

    default:
      return state;
  }
};

export default beltsReducer;
