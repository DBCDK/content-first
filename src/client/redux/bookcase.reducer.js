const defaultState = {
  celebs: [
    {
      id: '1',
      name: 'Helene Poulsen',
      img: 'img/bookcase/HP.jpg',
      description:
        'Helene Jeppesen er Danmarks mest fulgte booktuber. Hun læser bredt – både klassikere og helt ny skønlitteratur. Reolen bærer præg af hendes store interesse og kærlighed til det engelske sprog.',
      bookcase: 'img/bookcase/HP-bogreol.jpg',
      books: [
        {
          pid: '870970-basis:51704185',
          position: {x: 24.5, y: 12},
          description:
            'Romanen er skrevet af forfatteren til ”Stoner” og har mange af de samme kvaliteter. Her foregår handlingen blot i naturen, i det vilde vesten, hvor en gruppe mænd tager på Bisonjagt.'
        },
        {
          pid: '870970-basis:52921589',
          position: {x: 38.5, y: 12}
        },
        {
          pid: '870970-basis:51283376',
          position: {x: 58.5, y: 27}
        },
        {
          pid: '870970-basis:52911567',
          position: {x: 49.3, y: 61}
        },
        {
          pid: '870970-basis:53077145',
          position: {x: 28.3, y: 76}
        },
        {
          pid: '870970-basis:29554129',
          position: {x: 90, y: 61}
        }
      ]
    },
    {
      id: '2',
      name: 'B. S. Christiansen',
      img: 'img/bookcase/BS.jpg',
      description:
        'er en dansk forhenværende elitesoldat i Jægerkorpset og fordragsholder. Han er især kendt for programserien På afveje med forskellige kendte personer.',
      bookcase: 'img/bookcase/BS-bogreol.jpg',
      books: [
        {
          pid: '870970-basis:52530423',
          position: {x: 13.5, y: 40}
        },
        {
          pid: '870970-basis:53079202',
          position: {x: 21.8, y: 44}
        },
        {
          pid: '870970-basis:52038014',
          position: {x: 26, y: 46}
        },
        {
          pid: '870970-basis:23211629',
          position: {x: 36.5, y: 46}
        }
      ]
    },
    {
      id: '3',
      name: 'Dar Salim',
      img: 'img/bookcase/DS.jpg',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed at rhoncus enim, ut facilisis velit. Donec eget augue vel lorem iaculis.',
      bookcase: 'img/bookcase/DS-bogreol.jpg',
      books: [
        {
          pid: '870970-basis:26378176',
          position: {x: 58, y: 37}
        },
        {
          pid: '870970-basis:53052932',
          position: {x: 58, y: 46}
        },
        {
          pid: '870970-basis:52728843',
          position: {x: 64, y: 55}
        },
        {
          pid: '870970-basis:29158347',
          position: {x: 64, y: 68}
        },
        {
          pid: '870970-basis:27617050',
          position: {x: 60, y: 83}
        }
      ]
    }
  ]
};

export const ON_BOOK_REQUEST_TEST = 'ON_BOOK_REQUEST_TEST';

const bookcaseReducer = (state = defaultState, action) => {
  switch (action.type) {
    case ON_BOOK_REQUEST_TEST: {
      return Object.assign({}, {state});
    }

    default:
      return state;
  }
};

export default bookcaseReducer;
