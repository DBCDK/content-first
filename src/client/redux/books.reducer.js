const defaultState = {
  pids: [],
  isLoading: false,
  books: []
};

export const BOOKS_REQUEST = 'BOOKS_REQUEST';
export const BOOKS_RESPONSE = 'BOOKS_RESPONSE';

const booksReducer = (state = defaultState, action) => {
  switch (action.type) {
    case BOOKS_REQUEST:
      return Object.assign({}, state, {pids: [action.pids]}, {isLoading: true});
    case BOOKS_RESPONSE:
      return Object.assign(
        {},
        state,
        {books: action.response},
        {isLoading: false}
      );
    default:
      return state;
  }
};

export default booksReducer;
