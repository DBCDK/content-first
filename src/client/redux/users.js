/* eslint no-undefined:0 */
import Immutable from 'immutable';
import request from 'superagent';
import {createSelector} from 'reselect';

const defaultState = Immutable.fromJS({});

export const REQUEST_USER = 'REQUEST_USER';
export const RECEIVE_USER = 'RECEIVE_USER';

//
// Reducer
//
export function usersReducer(state = defaultState, action) {
  switch (action.type) {
    case REQUEST_USER:
      return state.get(action.id, false)
        ? state
        : state.setIn([action.id, 'loading'], true);

    case RECEIVE_USER:
      return state.set(action.id, Immutable.fromJS(action.user));

    default:
      return state;
  }
}

//
// Middleware
//
export const usersMiddleware = store => next => action => {
  switch (action.type) {
    case REQUEST_USER:
      if (!store.getState().users.get(action.id, false)) {
        (async () => {
          try {
            const response = await request.get(
              `/v1/user/${encodeURIComponent(action.id)}`
            );
            store.dispatch({
              type: RECEIVE_USER,
              id: action.id,
              user: response.body.data
            });
          } catch (e) {
            store.dispatch({
              type: 'LOG_ERROR',
              userId: action.id,
              error: String(e),
              msg: 'Error when fetching user'
            });

            // Clear user in state, such that we try to fetch it again, if we try again
            store.dispatch({
              type: RECEIVE_USER,
              id: action.id,
              user: undefined
            });
          }
        })();
      }
      return next(action);
    default:
      return next(action);
  }
};

// -------------SELECTORS-----------------
export const createGetUserSelector = () =>
  createSelector(
    [state => state.users, (state, {id}) => id],
    (users, id) => {
      const user = users.get(id);
      if (user) {
        return user.toJS();
      }
      return null;
    }
  );
export const getUser = createGetUserSelector();
export const createGetUsersSelector = () =>
  createSelector(
    state => state.users,
    users => {
      return users.toJS();
    }
  );
