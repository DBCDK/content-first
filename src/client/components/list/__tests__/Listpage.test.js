import React from 'react';
import {Provider} from 'react-redux';
import ListPage from '../ListPage.container';
import renderer from 'react-test-renderer';
import {createStore, combineReducers} from 'redux';
import listReducer, {addList} from '../../../redux/list.reducer';
import {usersReducer} from '../../../redux/users';

const reducer = (state, action) => {
  if (action.type === 'CLEAR_STATE') {
    return {
      listReducer: {
        lists: {}
      }
    };
  }
  return combineReducers({
    listReducer,
    users: usersReducer
  })(state, action);
};

const createTestElement = id => {
  return {
    book: {
      pid: 'pid' + id,
      title: 'some title ' + id,
      creator: 'some creator' + id
    },
    links: {
      cover: '/cover' + id
    },
    description: 'some description' + id
  };
};

const createList = id => {
  return {
    id: id,
    title: 'some list',
    description: 'some description',
    list: [createTestElement(1), createTestElement(2), createTestElement(3)]
  };
};

jest.mock(
  '../../work/WorkItemConnected.component.js',
  () => 'WorkItemConnectedMock'
);

jest.mock('../../comments/Comment.container.js', () => 'Comments');
jest.mock('../../list/templates/SimpleList.component.js', () => 'Comments');
jest.mock('../../list/templates/CircleTemplate.container.js', () => 'Comments');

describe('ListView', () => {
  test('List is rendered', () => {
    const store = createStore(reducer);
    store.dispatch(addList(createList('list1')));
    const tree = renderer
      .create(
        <Provider store={store}>
          <ListPage id="list1" />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('no list is rendered when id is invalid', () => {
    const store = createStore(reducer);
    const tree = renderer
      .create(
        <Provider store={store}>
          <ListPage id="list1" />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
