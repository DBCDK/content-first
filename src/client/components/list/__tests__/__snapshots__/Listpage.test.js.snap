// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListView List is rendered 1`] = `
<div
  className="list-wrapper tl"
>
  <h1
    className="t-title h-underline mb4"
  >
    some list
  </h1>
  <div
    className="row"
  >
    <div
      className="list tl col-xs-8"
    >
      <div
        className="simplelist"
      >
        <div
          className="row mb4"
        >
          <div
            className="col-xs-3 tc"
          >
            <div
              className="profile-card "
            >
              <span
                className="profile-image small round"
              >
                <img
                  alt="Profile Name"
                  className="cover"
                  src="http://p-hold.com/200/200"
                />
              </span>
              <p
                className="profile-name t-body h4"
              >
                Profile Name
              </p>
            </div>
          </div>
          <div
            className="col-xs-9"
          >
            <p
              className="t-body"
            >
              some description
            </p>
          </div>
        </div>
        <div
          className="list"
        >
          <div
            className="row simplelist-item mb4"
          >
            <div
              className="meta col-xs-3 tc"
            >
              <WorkItemConnectedMock
                showTaxonomy={false}
                work={
                  Object {
                    "book": Object {
                      "creator": "some creator1",
                      "pid": "pid1",
                      "title": "some title 1",
                    },
                  }
                }
                workClass="work simplelist"
              />
            </div>
            <div
              className="meta col-xs-9"
            >
              <h4
                className="w-title h-tight"
              >
                some title 1
              </h4>
              <h5
                className="w-creator h-tight mb2"
              >
                some creator1
              </h5>
              <div
                className="profile-description"
              >
                <div
                  className="profile-list mb1"
                >
                  <span
                    className="profile-image small round"
                  >
                    <img
                      alt="Profile Name"
                      className="cover"
                      src="http://p-hold.com/200/200"
                    />
                  </span>
                  <p
                    className="profile-name t-body h4"
                  >
                    Profile Name
                  </p>
                </div>
                <p
                  className="t-body"
                >
                  some description1
                </p>
              </div>
            </div>
          </div>
          <div
            className="row simplelist-item mb4"
          >
            <div
              className="meta col-xs-3 tc"
            >
              <WorkItemConnectedMock
                showTaxonomy={false}
                work={
                  Object {
                    "book": Object {
                      "creator": "some creator2",
                      "pid": "pid2",
                      "title": "some title 2",
                    },
                  }
                }
                workClass="work simplelist"
              />
            </div>
            <div
              className="meta col-xs-9"
            >
              <h4
                className="w-title h-tight"
              >
                some title 2
              </h4>
              <h5
                className="w-creator h-tight mb2"
              >
                some creator2
              </h5>
              <div
                className="profile-description"
              >
                <div
                  className="profile-list mb1"
                >
                  <span
                    className="profile-image small round"
                  >
                    <img
                      alt="Profile Name"
                      className="cover"
                      src="http://p-hold.com/200/200"
                    />
                  </span>
                  <p
                    className="profile-name t-body h4"
                  >
                    Profile Name
                  </p>
                </div>
                <p
                  className="t-body"
                >
                  some description2
                </p>
              </div>
            </div>
          </div>
          <div
            className="row simplelist-item mb4"
          >
            <div
              className="meta col-xs-3 tc"
            >
              <WorkItemConnectedMock
                showTaxonomy={false}
                work={
                  Object {
                    "book": Object {
                      "creator": "some creator3",
                      "pid": "pid3",
                      "title": "some title 3",
                    },
                  }
                }
                workClass="work simplelist"
              />
            </div>
            <div
              className="meta col-xs-9"
            >
              <h4
                className="w-title h-tight"
              >
                some title 3
              </h4>
              <h5
                className="w-creator h-tight mb2"
              >
                some creator3
              </h5>
              <div
                className="profile-description"
              >
                <div
                  className="profile-list mb1"
                >
                  <span
                    className="profile-image small round"
                  >
                    <img
                      alt="Profile Name"
                      className="cover"
                      src="http://p-hold.com/200/200"
                    />
                  </span>
                  <p
                    className="profile-name t-body h4"
                  >
                    Profile Name
                  </p>
                </div>
                <p
                  className="t-body"
                >
                  some description3
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      ;
    </div>
    <div
      className="col-xs-4"
    >
      <a
        className=""
        href="/lister/list1/rediger"
        onClick={[Function]}
      >
        Rediger liste
      </a>
    </div>
  </div>
</div>
`;

exports[`ListView no list is rendered when id is invalid 1`] = `
<div>
  Listen findes ikke
</div>
`;
